docs/source/developer_guides/custom_models.md

```diff
@@ -238,3 +238,64 @@ peft_model.print_trainable_parameters()
 ```python
 print(peft_model.targeted_module_names)
 ```
+
+## Unsupported module types
+
+In PEFT in general, methods like LoRA only work if the target modules are supported by PEFT. For example, it's possible to apply LoRA to `nn.Linear` and `nn.Conv2d` layers, but not, for instance, to `nn.LSTM`. If you find that the class of layer you want to apply PEFT to is not supported, feel free to open a [PEFT issue on GitHub](https://github.com/huggingface/peft/issues) and request the feature. Maintainers will implement this or guide you on how to implement this yourself if the demand for this module type is sufficiently high.
+
+### Experimental support for dynamic dispatch of custom modules in LoRA
+
+> [!WARNING]
+> This feature is experimental and subject to change, depending on its reception by the community. If we find that there is a big demand, we will introduce a public, stable API.
+
+For LoRA specifically, PEFT supports an experimental API for custom module types. Let's assume that you have a LoRA implementation for LSTMs. Normally, you would not be able to tell PEFT to use it, even if it would theoretically work with PEFT. With dynamic dispatch of custom layers, you can, however, make this work.
+
+The experimental API currently looks like this:
+
+```python
+class MyLoraLSTMLayer:
+    ...
+
+base_model = ...  # load the base model that uses LSTMs
+
+# add the LSTM layer names to target_modules
+config = LoraConfig(..., target_modules=["lstm"])
+# define a mapping from base layer type to LoRA layer type
+custom_module_mapping = {nn.LSTM: MyLoraLSTMLayer}
+# register the new mapping
+config._register_custom_module(custom_module_mapping)
+# after registration, create the PEFT model
+peft_model = get_peft_model(base_model, config)
+# do training
+```
+
+<Tip>
+When you call [`~get_peft_model`]`, you will see a warning because PEFT does not recognize the module type being targeted. In this case, you can ignore this warning.
+</Tip>
+
+By supplying a custom mapping, PEFT will take care of checking the layers of the base model against this custom mapping first and dispatching to the custom LoRA layer type if there is a match. Only after this will PEFT check the built-in LoRA layer types for a match. Therefore, this feature can also be used to override existing dispatch logic, e.g. if you want to use your own LoRA layer for `nn.Linear` instead of using the one provided by PEFT.
+
+When creating your custom LoRA module, please follow the same rules as the existing LoRA modules do. For this, check the [LoRA layer implementation](https://github.com/huggingface/peft/blob/main/src/peft/tuners/lora/layer.py). Notable constraints to consider:
+
+- The custom module should inherit from `nn.Module` and `peft.tuners.lora.layer.LoraLayer`
```

Comment:
here we should IMO state that the signature of the init method should have `base_layer` and  `adapter_name` in the correct order otherwise the API will fail

Reply:
Good point, I added an entry for the `__init__`.